game-app/game-core/src/main/java/games/strategy/triplea/ui/TechnologyPanel.java

```diff
@@ -0,0 +1,288 @@
+      getIcon(p);
+    }
+  }
+
+  static class JComponentTableCellRenderer implements TableCellRenderer {
+    @Override
+    public Component getTableCellRendererComponent(
+        final JTable table,
+        final Object value,
+        final boolean isSelected,
+        final boolean hasFocus,
+        final int row,
+        final int column) {
+      return (JComponent) value;
+    }
+  }
+
+  class TechTableModel extends AbstractTableModel {
+    private static final long serialVersionUID = -4612476336419396081L;
+    /* Flag to indicate whether data needs to be recalculated */
+    private boolean isDirty = true;
+    /* Column Header Names */
+    /* Row Header Names */
+    private String[] colList;
+    /* Underlying data for the table */
+    private final String[][] data;
+    /* Convenience mapping of country names -> col */
+    private final Map<String, Integer> colMap = new HashMap<>();
+    /* Convenience mapping of technology names -> row */
+    private final Map<String, Integer> rowMap = new HashMap<>();
+
+    TechTableModel() {
+      initColList();
+      /* Load the country -> col mapping */
+      for (int i = 0; i < colList.length; i++) {
+        colMap.put(colList[i], i + 1);
+      }
+      data = getDataAndInitRowMap();
+      clearAdvances();
+    }
+
+    private synchronized String[] @NotNull [] getDataAndInitRowMap() {
+      final String[][] dataTable;
+      boolean useTech = false;
+      // copy so that the object doesn't change underneath us
+      final GameData gameDataSync = TechnologyPanel.this.gameData;
+      try (GameData.Unlocker ignored = gameDataSync.acquireReadLock()) {
+        final int numTechs =
+            TechAdvance.getTechAdvances(gameDataSync.getTechnologyFrontier()).size();
+        if (gameDataSync.getResourceList().getResource(Constants.TECH_TOKENS) != null) {
+          useTech = true;
+          dataTable = new String[numTechs + 1][colList.length + 2];
+        } else {
+          dataTable = new String[numTechs][colList.length + 1];
+        }
+      }
+      /* Load the technology -> row mapping */
+      int row = 0;
+      if (useTech) {
+        rowMap.put("Tokens", row);
+        dataTable[row][0] = "Tokens";
+        row++;
+      }
+      final List<TechAdvance> techAdvances =
+          TechAdvance.getTechAdvances(gameDataSync.getTechnologyFrontier(), null);
+      for (final TechAdvance tech : techAdvances) {
+        rowMap.put(tech.getName(), row);
+        dataTable[row][0] = tech.getName();
+        row++;
+      }
+      return dataTable;
+    }
+
+    private void clearAdvances() {
+      /* Initialize the table with the tech names */
+      for (int i = 0; i < data.length; i++) {
+        for (int j = 1; j <= colList.length; j++) {
+          data[i][j] = "";
+        }
+      }
+    }
+
+    private void initColList() {
+      final List<GamePlayer> players = gameData.getPlayerList().getPlayers();
+      colList = new String[players.size()];
```

Comment:
`colList = (String[]) players.stream().map(GamePlayer::getName).toArray();` should not require the loop than and especially avoid `get(i)` for each entry of `players`.